packages/util-endpoints/src/cache/EndpointCache.ts

```diff
@@ -0,0 +1,62 @@
+import type { EndpointParams, EndpointV2 } from "@smithy/types";
+
+/**
+ * @internal
+ *
+ * Cache for endpoint ruleSet resolution.
+ */
+export class EndpointCache {
+  private capacity: number;
+  private data = new Map<string, EndpointV2>();
+  private parameters: string[] = [];
+
+  /**
+   * @param [params] - list of params to consider as part of the cache key.
+   *
+   * If the params list is not populated, all object keys will be considered.
+   * This may be out of order depending on how the object is created and arrives to this class.
+   */
+  public constructor({ size, params }: { size?: number; params?: string[] }) {
+    this.capacity = size ?? 50;
+    if (params) {
+      this.parameters = params;
+    }
+  }
+
+  /**
+   * @param endpointParams - query for endpoint.
+   * @param resolver - provider of the value if not present.
+   * @returns endpoint corresponding to the query.
+   */
+  public get(endpointParams: EndpointParams, resolver: () => EndpointV2): EndpointV2 {
+    const key = this.hash(endpointParams);
+    if (!this.data.has(key)) {
+      if (this.data.size > this.capacity + 10) {
+        const keys = this.data.keys();
+        let i = 0;
+        while (true) {
+          const { value, done } = keys.next();
+          this.data.delete(value);
+          if (done || ++i > 10) {
+            break;
+          }
+        }
+      }
+      this.data.set(key, resolver());
+    }
+    return this.data.get(key)!;
+  }
+
+  public size() {
+    return this.data.size;
+  }
+
+  private hash(endpointParams: EndpointParams): string {
+    let buffer = "";
+    const params = this.parameters.length ? this.parameters : Object.keys(endpointParams);
+    for (const param of params) {
+      buffer += endpointParams[param] ?? "" + "|";
```

Comment:
Some questions about this hashing method:
1. Does the specific `param` not matter for the hash? At least in the case where `params` is going to be `Object.keys(endpointParams)`, seems like it could matter
2. Can `endpointParams[param]` contain a `|` that would mess with the hash? 
3. Should anything be appended in the case where the key `param` doesn't exist? Might matter in the same case as 1.

Maybe these are edge cases that don't sufficiently impact the effectiveness of the hash though.

Reply:
1. params needs to be consistently sorted, so parameters are alphabetized by codegen. In the case where Object.keys are used, this is best-effort since they are not guaranteed to be in any order.
2. a parameter with a delimiter in it can affect the hash, yes. `A="|",B=void` and `A=void,B="|"` would both hash to `|||`. I'm going to make the delimiter `|;` and abandon caching if this sequence exists in a value.
3. A delimiter as placeholder is enough in this case.